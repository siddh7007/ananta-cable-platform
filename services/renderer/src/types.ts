/**
 * RenderDSL - Minimal drawing specification language for cable assemblies
 */

export interface RenderDSL {
  /** Assembly metadata */
  meta: {
    assembly_id: string;
    schema_hash: string;
    created_at?: string;
  };

  /** Physical dimensions */
  dimensions: {
    oal_mm: number;
    tolerance_mm: number;
    broken_dim?: boolean; // Use broken dimension style
  };

  /** Cable/ribbon specification */
  cable: RibbonCable | RoundCable;

  /** End A connector */
  endA: ConnectorFootprint;

  /** End B connector */
  endB: ConnectorFootprint;

  /** Wire nets */
  nets: Net[];

  /** Label/callout placements */
  labels: Label[];

  /** Notes pack ID */
  notesPack: string;

  /** QR code URL */
  qr?: string;
}

export type RibbonCable = {
  type: 'ribbon';
  ways: number;
  pitch_in: number;
  red_stripe?: boolean;
};

export type RoundCable = {
  type: 'round';
  conductors: number;
  awg?: number;
  shield?: 'none' | 'foil' | 'braid' | 'foil_braid';
};

export interface ConnectorFootprint {
  connector_mpn: string;
  type: 'idc' | 'crimp' | 'solder' | 'ring_lug';
  positions: number;
  orientation?: 'horizontal' | 'vertical';
  pin_numbering?: 'sequential' | 'dual_row';
}

export interface Net {
  circuit: string;
  endA_pin: string;
  endB_pin: string;
  color?: string;
  shield?: 'none' | 'fold_back' | 'isolated' | 'pigtail';
}

export interface Label {
  text: string;
  anchor: 'endA' | 'endB' | 'cable' | 'dimension';
  offset_x?: number;
  offset_y?: number;
}

/**
 * Render request payload
 */
export interface RenderRequest {
  dsl: RenderDSL;
  templatePackId: string;
  format?: 'svg' | 'pdf' | 'png';
}

/**
 * Render response
 */
export interface RenderResponse {
  svg?: string;
  pdf?: string; // base64
  png?: string; // base64
  manifest: RenderManifest;
}

export interface RenderManifest {
  rendererVersion: string;
  templatePackId: string;
  rendererKind: 'svg2d';
  schemaHash: string;
}

/**
 * Template pack manifest
 */
export interface TemplatePackManifest {
  id: string;
  version: string;
  name: string;
  paper: 'A3' | 'Letter' | 'A4';
  dimensions: {
    width_mm: number;
    height_mm: number;
  };
  margins: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  };
  styles: {
    lineWidth: number;
    fontSize: number;
    font: string;
    colors: {
      primary: string;
      secondary: string;
      accent: string;
    };
  };
}

/**
 * Internal rendering context
 */
export interface RenderContext {
  dsl: RenderDSL;
  template: TemplatePackManifest;
  viewport: {
    width: number;
    height: number;
    contentX: number;
    contentY: number;
    contentWidth: number;
    contentHeight: number;
  };
  topology: TopologyResult;
  routing: RoutingResult;
  collisions: CollisionResult;
  dimensions: DimensionResult;
}

export interface TopologyResult {
  endA: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
  endB: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
  cableRegion: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
}

export interface RoutingResult {
  paths: NetPath[];
  ribbonLayout?: {
    startX: number;
    startY: number;
    endX: number;
    endY: number;
    lanes: number;
    pitch: number;
  };
}

export interface NetPath {
  circuit: string;
  lane: number;
  points: { x: number; y: number }[];
  color?: string;
}

export interface CollisionResult {
  adjustedLabels: Array<{
    original: Label;
    x: number;
    y: number;
  }>;
}

export interface DimensionResult {
  oal: {
    x1: number;
    y1: number;
    x2: number;
    y2: number;
    value_mm: number;
    tolerance_mm: number;
    broken: boolean;
  };
}

/**
 * SVG generation options
 */
export interface SVGOptions {
  precision?: number; // Decimal places (default: 2)
  prettyPrint?: boolean;
  includeMetadata?: boolean;
}
