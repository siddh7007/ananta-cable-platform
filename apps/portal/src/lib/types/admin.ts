// Admin types for the Cable Platform portal
// These types support admin functionality like user management

export interface AdminUser {
  id: string;
  name: string;
  email: string;
  roles: string[];
  status: 'active' | 'deactivated';
  createdAt?: string;
  updatedAt?: string;
}

export interface AdminUsersResponse {
  users: AdminUser[];
  total: number;
  limit: number;
  offset: number;
}

export interface AdminUserActionResponse {
  success: boolean;
  user: AdminUser;
  message?: string;
}

// Database statistics types
export interface DbConnectionStatus {
  status: 'ok' | 'fail' | 'unknown';
  latencyMs?: number;
  note?: string;
}

export interface DbStats {
  supabase: DbConnectionStatus;
  pgExtra?: DbConnectionStatus;
  oracle?: DbConnectionStatus;
  counts: {
    workspaces: number;
    projects: number;
    boms: number;
    orders: number;
    users: number;
  };
  ts: string; // ISO timestamp
}

export interface AdminDbData {
  stats?: DbStats;
  error?: string;
}
