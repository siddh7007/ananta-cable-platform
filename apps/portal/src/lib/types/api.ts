// API types for the Cable Platform frontend
// These should eventually be generated from OpenAPI schemas

// DRC types (matching OpenAPI schemas from packages/contracts/openapi.yaml)
export type DrcSeverity = 'info' | 'warning' | 'error';
export type DrcStatus = 'pass' | 'warning' | 'error';
export type DrcDomain = 'mechanical' | 'electrical' | 'standards' | 'labeling' | 'consistency';
export type DrcFixEffect = 'non_destructive' | 'substitution' | 're_synthesis_required';

export interface DrcFinding {
  id: string;
  severity: DrcSeverity;
  domain: DrcDomain;
  code: string;
  message: string;
  where?: string;
  refs?: string[];
}

export interface DrcFix {
  id: string;
  label: string;
  description: string;
  applies_to: string[]; // Array of finding IDs
  effect: DrcFixEffect;
}

export interface DrcReport {
  assembly_id: string;
  ruleset_id: string;
  version: string;
  passed: boolean;
  errors: number;
  warnings: number;
  findings: DrcFinding[];
  fixes: DrcFix[];
  generated_at: string;
}

// Legacy/simplified types for backward compatibility
export interface DrcIssue {
  type: string;
  severity: DrcSeverity;
  message: string;
  location: string;
  suggestion?: string;
}

export interface DrcResult {
  status: DrcStatus;
  issues: DrcIssue[];
  summary: string;
}

// Legacy aliases for backward compatibility
export type DRCFinding = DrcIssue;
export type DRCReport = DrcResult;

// Connector types
export interface ConnectorSummary {
  mpn: string;
  manufacturer?: string;
  positions?: number;
  type?: string;
}

export interface ConnectorMetadata {
  mpn: string;
  manufacturer: string;
  series: string;
  positions: number;
  pitch_mm?: number;
  current_rating_a?: number;
  voltage_rating_v?: number;
}

export type TerminationType = 'crimp' | 'solder' | 'idc' | 'wire_to_board';

// Assembly types
export interface Endpoint {
  connector: ConnectorSummary;
  termination: TerminationType;
  label?: string;
}

export interface AssemblyStep1 {
  id: string;
  name: string;
  endpoints: Record<string, Endpoint>;
  conductor_count?: number;
  conductor_awg?: number;
}

export interface SynthesisProposal {
  assembly_id: string;
  cable_type: string;
  cable_length_mm: number;
  conductor_awg: number;
  conductor_count: number;
  shield_type?: string;
  endpoints: Record<string, Endpoint>;
}

// Region type
export interface Region {
  code: string;
  name: string;
  locale: string;
}