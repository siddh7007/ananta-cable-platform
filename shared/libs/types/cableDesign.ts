import { z } from "zod";

/**
 * Zod schema mirroring shared/contracts/schemas/v1/cable-design.schema.json
 */
export const CableDesignSchema = z.object({
  id: z.string(),
  name: z.string(),
  cores: z.number().int().min(1),
});

export type CableDesign = z.infer<typeof CableDesignSchema>;
