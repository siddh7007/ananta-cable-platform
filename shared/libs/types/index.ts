export * from "./cableDesign";
export * from "./drcResult";
export * from "./vendor";
