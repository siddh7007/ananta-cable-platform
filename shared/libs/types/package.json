{
  "name": "@cable-platform/types",
  "version": "0.0.0",
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "scripts": {
    "build": "tsc -p tsconfig.json"
  },
  "dependencies": {
    "zod": "^3.25.3"
  }
}
