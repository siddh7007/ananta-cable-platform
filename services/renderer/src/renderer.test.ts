import { describe, it, expect, beforeAll } from 'vitest';
import { renderToSVG } from './renderer/index.js';
import { RenderDSL } from './types.js';
import { mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';

// Setup mock template pack
beforeAll(() => {
  const templateDir = join(process.cwd(), 'packages/templates/basic-a3');
  mkdirSync(templateDir, { recursive: true });
  
  const manifest = {
    id: 'basic-a3',
    version: '1.0.0',
    name: 'Basic A3 Template',
    paper: 'A3',
    dimensions: {
      width_mm: 420,
      height_mm: 297,
    },
    margins: {
      top: 20,
      right: 20,
      bottom: 30,
      left: 20,
    },
    styles: {
      lineWidth: 0.35,
      fontSize: 3.5,
      font: 'DIN-Regular',
      colors: {
        primary: '#000000',
        secondary: '#666666',
        accent: '#FF0000',
      },
    },
  };
  
  writeFileSync(
    join(templateDir, 'manifest.json'),
    JSON.stringify(manifest, null, 2)
  );
});

describe('Renderer', () => {
  describe('Ribbon Cable with IDC', () => {
    it('renders 12-way ribbon with red stripe and pin-1 indicators', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_ribbon_12way',
          schema_hash: 'hash123abc',
        },
        dimensions: {
          oal_mm: 1250,
          tolerance_mm: 5,
        },
        cable: {
          type: 'ribbon',
          ways: 12,
          pitch_in: 0.05,
          red_stripe: true,
        },
        endA: {
          connector_mpn: 'TE-102345',
          type: 'idc',
          positions: 12,
        },
        endB: {
          connector_mpn: 'TE-102345',
          type: 'idc',
          positions: 12,
        },
        nets: Array.from({ length: 12 }, (_, i) => ({
          circuit: `D${i}`,
          endA_pin: String(i + 1),
          endB_pin: String(i + 1),
          color: ['brown', 'red', 'orange', 'yellow', 'green', 'blue', 'violet', 'gray', 'white', 'black', 'pink', 'turquoise'][i],
        })),
        labels: [
          {
            text: 'RIBBON CABLE ASSY',
            anchor: 'cable',
          },
        ],
        notesPack: 'IPC-620-CLASS-2',
        qr: 'https://example.com/asm_ribbon_12way',
      };

      const svg = await renderToSVG(dsl, 'basic-a3');

      // Structural assertions
      expect(svg).toContain('<?xml version="1.0"');
      expect(svg).toContain('<svg');
      expect(svg).toContain('assembly id="asm_ribbon_12way"');
      expect(svg).toContain('template id="basic-a3"');
      
      // Pin-1 indicators (both ends)
      expect(svg).toContain('id="asm_ribbon_12way-pin1-L"');
      expect(svg).toContain('id="asm_ribbon_12way-pin1-R"');
      
      // Red stripe
      expect(svg).toContain('red-stripe');
      
      // Connectors
      expect(svg).toContain('id="asm_ribbon_12way-endA"');
      expect(svg).toContain('id="asm_ribbon_12way-endB"');
      
      // All 12 nets
      for (let i = 0; i < 12; i++) {
        expect(svg).toContain(`id="asm_ribbon_12way-net-D${i}"`);
      }
      
      // Dimension
      expect(svg).toContain('id="asm_ribbon_12way-dim-oal"');
      expect(svg).toContain('1250 ±5 mm');
      
      // Label
      expect(svg).toContain('RIBBON CABLE ASSY');
      
      // Notes
      expect(svg).toContain('IPC-620-CLASS-2');
      
      // QR code
      expect(svg).toContain('id="asm_ribbon_12way-qr"');

      // Snapshot test for determinism
      expect(svg).toMatchSnapshot();
    });

    it('counts correct number of elements', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_count_test',
          schema_hash: 'hash456def',
        },
        dimensions: {
          oal_mm: 500,
          tolerance_mm: 3,
        },
        cable: {
          type: 'ribbon',
          ways: 6,
          pitch_in: 0.05,
        },
        endA: {
          connector_mpn: 'TE-102345',
          type: 'idc',
          positions: 6,
        },
        endB: {
          connector_mpn: 'TE-102345',
          type: 'idc',
          positions: 6,
        },
        nets: Array.from({ length: 6 }, (_, i) => ({
          circuit: `N${i}`,
          endA_pin: String(i + 1),
          endB_pin: String(i + 1),
        })),
        labels: [],
        notesPack: 'STD',
      };

      const svg = await renderToSVG(dsl, 'basic-a3');

      // Count elements
      const connectorCount = (svg.match(/class="connector"/g) || []).length;
      const wireCount = (svg.match(/class="wire"/g) || []).length;
      const pin1Count = (svg.match(/id="[^"]*-pin1-[LR]"/g) || []).length;
      const dimCount = (svg.match(/id="[^"]*-dim-/g) || []).length;

      expect(connectorCount).toBe(2); // endA and endB
      expect(wireCount).toBe(6); // 6 nets
      expect(pin1Count).toBe(2); // pin-1 indicators on both ends
      expect(dimCount).toBe(1); // OAL dimension
    });
  });

  describe('Round Cable with Labels', () => {
    it('renders 2-conductor power cable with labels and dimension', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_power_2cond',
          schema_hash: 'hash789ghi',
        },
        dimensions: {
          oal_mm: 800,
          tolerance_mm: 10,
        },
        cable: {
          type: 'round',
          conductors: 2,
          awg: 18,
          shield: 'none',
        },
        endA: {
          connector_mpn: 'PHOENIX-123',
          type: 'crimp',
          positions: 2,
        },
        endB: {
          connector_mpn: 'PHOENIX-123',
          type: 'crimp',
          positions: 2,
        },
        nets: [
          {
            circuit: '+48V',
            endA_pin: '1',
            endB_pin: '1',
            color: 'red',
          },
          {
            circuit: 'RTN',
            endA_pin: '2',
            endB_pin: '2',
            color: 'black',
          },
        ],
        labels: [
          {
            text: '+48V POWER',
            anchor: 'endA',
            offset_y: -5,
          },
          {
            text: 'LOAD',
            anchor: 'endB',
            offset_y: -5,
          },
        ],
        notesPack: 'UL-LISTED',
        qr: 'https://example.com/asm_power_2cond',
      };

      const svg = await renderToSVG(dsl, 'basic-a3');

      // Structural assertions
      expect(svg).toContain('assembly id="asm_power_2cond"');
      
      // Both nets
      expect(svg).toContain('id="asm_power_2cond-net-+48V"');
      expect(svg).toContain('id="asm_power_2cond-net-RTN"');
      
      // Wire colors
      expect(svg).toContain('stroke="red"');
      expect(svg).toContain('stroke="black"');
      
      // Labels
      expect(svg).toContain('+48V POWER');
      expect(svg).toContain('LOAD');
      
      // Dimension
      expect(svg).toContain('800 ±10 mm');
      
      // Notes
      expect(svg).toContain('UL-LISTED');

      // Snapshot test
      expect(svg).toMatchSnapshot();
    });
  });

  describe('Broken Dimension', () => {
    it('renders broken dimension glyph when needed', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_broken_dim',
          schema_hash: 'hash_broken',
        },
        dimensions: {
          oal_mm: 2500,
          tolerance_mm: 15,
          broken_dim: true,
        },
        cable: {
          type: 'round',
          conductors: 4,
        },
        endA: {
          connector_mpn: 'CONN-A',
          type: 'crimp',
          positions: 4,
        },
        endB: {
          connector_mpn: 'CONN-B',
          type: 'crimp',
          positions: 4,
        },
        nets: [
          { circuit: 'A', endA_pin: '1', endB_pin: '1' },
          { circuit: 'B', endA_pin: '2', endB_pin: '2' },
          { circuit: 'C', endA_pin: '3', endB_pin: '3' },
          { circuit: 'D', endA_pin: '4', endB_pin: '4' },
        ],
        labels: [],
        notesPack: 'STD',
      };

      const svg = await renderToSVG(dsl, 'basic-a3');

      // Check for broken dimension mark (zigzag path)
      expect(svg).toContain('<path d="M');
      expect(svg).toContain('2500 ±15 mm');

      // Snapshot test
      expect(svg).toMatchSnapshot();
    });
  });

  describe('Determinism', () => {
    it('produces identical SVG across multiple renders', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_determinism_test',
          schema_hash: 'hash_det',
        },
        dimensions: {
          oal_mm: 1000,
          tolerance_mm: 5,
        },
        cable: {
          type: 'ribbon',
          ways: 4,
          pitch_in: 0.05,
        },
        endA: {
          connector_mpn: 'TE-X',
          type: 'idc',
          positions: 4,
        },
        endB: {
          connector_mpn: 'TE-Y',
          type: 'idc',
          positions: 4,
        },
        nets: [
          { circuit: 'A', endA_pin: '1', endB_pin: '1' },
          { circuit: 'B', endA_pin: '2', endB_pin: '2' },
          { circuit: 'C', endA_pin: '3', endB_pin: '3' },
          { circuit: 'D', endA_pin: '4', endB_pin: '4' },
        ],
        labels: [],
        notesPack: 'STD',
      };

      // Render multiple times
      const svg1 = await renderToSVG(dsl, 'basic-a3');
      const svg2 = await renderToSVG(dsl, 'basic-a3');
      const svg3 = await renderToSVG(dsl, 'basic-a3');

      // All should be identical
      expect(svg1).toBe(svg2);
      expect(svg2).toBe(svg3);
    });

    it('normalizes numeric values to 2 decimals', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_precision',
          schema_hash: 'hash_prec',
        },
        dimensions: {
          oal_mm: 333.333333,
          tolerance_mm: 2.777777,
        },
        cable: {
          type: 'round',
          conductors: 1,
        },
        endA: {
          connector_mpn: 'C1',
          type: 'solder',
          positions: 1,
        },
        endB: {
          connector_mpn: 'C2',
          type: 'solder',
          positions: 1,
        },
        nets: [{ circuit: 'X', endA_pin: '1', endB_pin: '1' }],
        labels: [],
        notesPack: 'STD',
      };

      const svg = await renderToSVG(dsl, 'basic-a3');

      // Check that coordinates are normalized to 2 decimals
      // Should not contain more than 2 decimal places
      const coordRegex = /[xy]1?="(\d+\.\d+)"/g;
      let match;
      const coords: number[] = [];
      
      while ((match = coordRegex.exec(svg)) !== null) {
        coords.push(parseFloat(match[1]));
      }

      coords.forEach(coord => {
        const decimalPlaces = (coord.toString().split('.')[1] || '').length;
        expect(decimalPlaces).toBeLessThanOrEqual(2);
      });
    });
  });

  describe('Template Pack Switching', () => {
    it('renders with STD-A3-IPC620 template pack', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_a3_test',
          schema_hash: 'hash_a3',
        },
        dimensions: {
          oal_mm: 1000,
          tolerance_mm: 5,
        },
        cable: {
          type: 'ribbon',
          ways: 8,
          pitch_in: 0.05,
          red_stripe: true,
        },
        endA: {
          connector_mpn: 'TEST-A',
          type: 'idc',
          positions: 8,
        },
        endB: {
          connector_mpn: 'TEST-B',
          type: 'idc',
          positions: 8,
        },
        nets: Array.from({ length: 8 }, (_, i) => ({
          circuit: `D${i}`,
          endA_pin: String(i + 1),
          endB_pin: String(i + 1),
        })),
        labels: [],
        notesPack: 'IPC-620-CLASS-2',
      };

      const svg = await renderToSVG(dsl, 'STD-A3-IPC620');

      expect(svg).toContain('width="420mm"');
      expect(svg).toContain('height="297mm"');
      expect(svg).toContain('STD-A3-IPC620');
      expect(svg).toContain('1.1.0');
      expect(svg).toContain('titleblock');
      expect(svg).toContain('notes-table');
      
      // Snapshot for visual comparison
      expect(svg).toMatchSnapshot('a3-template');
    });

    it('renders with STD-Letter-IPC620 template pack', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_letter_test',
          schema_hash: 'hash_letter',
        },
        dimensions: {
          oal_mm: 800,
          tolerance_mm: 5,
        },
        cable: {
          type: 'ribbon',
          ways: 8,
          pitch_in: 0.05,
          red_stripe: true,
        },
        endA: {
          connector_mpn: 'TEST-A',
          type: 'idc',
          positions: 8,
        },
        endB: {
          connector_mpn: 'TEST-B',
          type: 'idc',
          positions: 8,
        },
        nets: Array.from({ length: 8 }, (_, i) => ({
          circuit: `D${i}`,
          endA_pin: String(i + 1),
          endB_pin: String(i + 1),
        })),
        labels: [],
        notesPack: 'IPC-620-CLASS-2',
      };

      const svg = await renderToSVG(dsl, 'STD-Letter-IPC620');

      expect(svg).toContain('width="279.4mm"');
      expect(svg).toContain('height="215.9mm"');
      expect(svg).toContain('STD-Letter-IPC620');
      expect(svg).toContain('1.0.0');
      expect(svg).toContain('titleblock');
      expect(svg).toContain('notes-table');
      
      // Snapshot for visual comparison
      expect(svg).toMatchSnapshot('letter-template');
    });

    it('snapshots differ only in expected regions (sheet size and titleblock)', async () => {
      const dsl: RenderDSL = {
        meta: {
          assembly_id: 'asm_compare',
          schema_hash: 'hash_compare',
        },
        dimensions: {
          oal_mm: 1000,
          tolerance_mm: 5,
        },
        cable: {
          type: 'ribbon',
          ways: 8,
          pitch_in: 0.05,
          red_stripe: false,
        },
        endA: {
          connector_mpn: 'TEST',
          type: 'idc',
          positions: 8,
        },
        endB: {
          connector_mpn: 'TEST',
          type: 'idc',
          positions: 8,
        },
        nets: Array.from({ length: 8 }, (_, i) => ({
          circuit: `D${i}`,
          endA_pin: String(i + 1),
          endB_pin: String(i + 1),
        })),
        labels: [],
        notesPack: 'IPC-620',
      };

      const svgA3 = await renderToSVG(dsl, 'STD-A3-IPC620');
      const svgLetter = await renderToSVG(dsl, 'STD-Letter-IPC620');

      // Both should contain the same assembly content
      expect(svgA3).toContain('asm_compare');
      expect(svgLetter).toContain('asm_compare');

      // But different dimensions
      expect(svgA3).toContain('420');
      expect(svgLetter).toContain('279.4');

      // Different titleblock positions (visual verification needed)
      expect(svgA3).not.toEqual(svgLetter);
    });
  });
});
