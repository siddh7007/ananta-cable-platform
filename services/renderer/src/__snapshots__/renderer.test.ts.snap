// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renderer > Broken Dimension > renders broken dimension glyph when needed 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="420mm" height="297mm" viewBox="0 0 420 297">
  <metadata>
    <assembly id="asm_broken_dim" schema="hash_broken"/>
    <template id="basic-a3" version="1.0.0"/>
  </metadata>
  <defs>
    <marker id="pin1-tri" markerWidth="4" markerHeight="4" refX="2" refY="2" orient="auto">
      <path d="M 0,0 L 4,2 L 0,4 Z" fill="black"/>
    </marker>
  </defs>
  <style>
    .connector { fill: none; stroke: black; stroke-width: 0.35; }
    .wire { fill: none; stroke-width: 0.25; }
    .dimension { fill: none; stroke: black; stroke-width: 0.2; }
    .text { font-family: DIN-Regular; font-size: 3.5px; fill: black; }
    .label { font-size: 2.8000000000000003px; }
  </style>
  <!-- Connectors -->
  <rect id="asm_broken_dim-endA" class="connector" x="60.00" y="136.00" width="20.00" height="15.00"/>
  <rect id="asm_broken_dim-endB" class="connector" x="340.00" y="136.00" width="20.00" height="15.00"/>
  <!-- Pin-1 Indicators -->
  <circle id="asm_broken_dim-pin1-L" cx="57.00" cy="138.00" r="1" fill="black"/>
  <circle id="asm_broken_dim-pin1-R" cx="363.00" cy="138.00" r="1" fill="black"/>
  <!-- Wires -->
  <path id="asm_broken_dim-net-A" class="wire" d="M 80.00,131.50 L 340.00,131.50" stroke="black"/>
  <path id="asm_broken_dim-net-B" class="wire" d="M 80.00,139.50 L 340.00,139.50" stroke="black"/>
  <path id="asm_broken_dim-net-C" class="wire" d="M 80.00,147.50 L 340.00,147.50" stroke="black"/>
  <path id="asm_broken_dim-net-D" class="wire" d="M 80.00,155.50 L 340.00,155.50" stroke="black"/>
  <!-- Dimensions -->
  <g id="asm_broken_dim-dim-oal" class="dimension">
    <line x1="60.00" y1="178.50" x2="205.00" y2="178.50"/>
    <line x1="215.00" y1="178.50" x2="360.00" y2="178.50"/>
    <path d="M 205.00,176.50 L 210.00,180.50 L 215.00,176.50" fill="none" stroke="black" stroke-width="0.2"/>
    <line x1="60.00" y1="173.50" x2="60.00" y2="180.50"/>
    <line x1="360.00" y1="173.50" x2="360.00" y2="180.50"/>
    <text x="210.00" y="176.50" class="text" text-anchor="middle">2500 ±15 mm</text>
  </g>
  <!-- Notes -->
  <text x="20.00" y="287.00" class="text label">Notes: STD</text>
</svg>"
`;

exports[`Renderer > Ribbon Cable with IDC > renders 12-way ribbon with red stripe and pin-1 indicators 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="420mm" height="297mm" viewBox="0 0 420 297">
  <metadata>
    <assembly id="asm_ribbon_12way" schema="hash123abc"/>
    <template id="basic-a3" version="1.0.0"/>
  </metadata>
  <defs>
    <marker id="pin1-tri" markerWidth="4" markerHeight="4" refX="2" refY="2" orient="auto">
      <path d="M 0,0 L 4,2 L 0,4 Z" fill="black"/>
    </marker>
    <pattern id="red-stripe" width="2" height="2" patternUnits="userSpaceOnUse">
      <rect width="1" height="2" fill="red"/>
    </pattern>
  </defs>
  <style>
    .connector { fill: none; stroke: black; stroke-width: 0.35; }
    .wire { fill: none; stroke-width: 0.25; }
    .dimension { fill: none; stroke: black; stroke-width: 0.2; }
    .text { font-family: DIN-Regular; font-size: 3.5px; fill: black; }
    .label { font-size: 2.8000000000000003px; }
  </style>
  <!-- Connectors -->
  <rect id="asm_ribbon_12way-endA" class="connector" x="60.00" y="136.00" width="20.00" height="15.00"/>
  <rect id="asm_ribbon_12way-endB" class="connector" x="340.00" y="136.00" width="20.00" height="15.00"/>
  <!-- Pin-1 Indicators -->
  <circle id="asm_ribbon_12way-pin1-L" cx="57.00" cy="138.00" r="1" fill="black"/>
  <circle id="asm_ribbon_12way-pin1-R" cx="363.00" cy="138.00" r="1" fill="black"/>
  <!-- Wires -->
  <path id="asm_ribbon_12way-net-D0" class="wire" d="M 80.00,136.51 L 340.00,136.51" stroke="brown"/>
  <path id="asm_ribbon_12way-net-D1" class="wire" d="M 80.00,137.78 L 340.00,137.78" stroke="red"/>
  <path id="asm_ribbon_12way-net-D2" class="wire" d="M 80.00,139.05 L 340.00,139.05" stroke="orange"/>
  <path id="asm_ribbon_12way-net-D3" class="wire" d="M 80.00,140.32 L 340.00,140.32" stroke="yellow"/>
  <path id="asm_ribbon_12way-net-D4" class="wire" d="M 80.00,141.59 L 340.00,141.59" stroke="green"/>
  <path id="asm_ribbon_12way-net-D5" class="wire" d="M 80.00,142.86 L 340.00,142.86" stroke="blue"/>
  <path id="asm_ribbon_12way-net-D6" class="wire" d="M 80.00,144.13 L 340.00,144.13" stroke="violet"/>
  <path id="asm_ribbon_12way-net-D7" class="wire" d="M 80.00,145.40 L 340.00,145.40" stroke="gray"/>
  <path id="asm_ribbon_12way-net-D8" class="wire" d="M 80.00,146.67 L 340.00,146.67" stroke="white"/>
  <path id="asm_ribbon_12way-net-D9" class="wire" d="M 80.00,147.94 L 340.00,147.94" stroke="black"/>
  <path id="asm_ribbon_12way-net-D10" class="wire" d="M 80.00,149.21 L 340.00,149.21" stroke="pink"/>
  <path id="asm_ribbon_12way-net-D11" class="wire" d="M 80.00,150.48 L 340.00,150.48" stroke="turquoise"/>
  <!-- Red Stripe -->
  <rect id="asm_ribbon_12way-red-stripe" x="80.00" y="136.51" width="260.00" height="1" fill="url(#red-stripe)"/>
  <!-- Dimensions -->
  <g id="asm_ribbon_12way-dim-oal" class="dimension">
    <line x1="60.00" y1="178.50" x2="360.00" y2="178.50"/>
    <line x1="60.00" y1="173.50" x2="60.00" y2="180.50"/>
    <line x1="360.00" y1="173.50" x2="360.00" y2="180.50"/>
    <text x="210.00" y="176.50" class="text" text-anchor="middle">1250 ±5 mm</text>
  </g>
  <!-- Labels -->
  <text id="asm_ribbon_12way-label-0" x="210.00" y="113.50" class="text label">RIBBON CABLE ASSY</text>
  <!-- Notes -->
  <text x="20.00" y="287.00" class="text label">Notes: IPC-620-CLASS-2</text>
  <!-- QR Code -->
  <rect id="asm_ribbon_12way-qr" x="390.00" y="267.00" width="20" height="20" fill="white" stroke="black" stroke-width="0.2"/>
  <text x="400.00" y="279.00" class="text label" text-anchor="middle" font-size="2">QR</text>
</svg>"
`;

exports[`Renderer > Round Cable with Labels > renders 2-conductor power cable with labels and dimension 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="420mm" height="297mm" viewBox="0 0 420 297">
  <metadata>
    <assembly id="asm_power_2cond" schema="hash789ghi"/>
    <template id="basic-a3" version="1.0.0"/>
  </metadata>
  <defs>
    <marker id="pin1-tri" markerWidth="4" markerHeight="4" refX="2" refY="2" orient="auto">
      <path d="M 0,0 L 4,2 L 0,4 Z" fill="black"/>
    </marker>
  </defs>
  <style>
    .connector { fill: none; stroke: black; stroke-width: 0.35; }
    .wire { fill: none; stroke-width: 0.25; }
    .dimension { fill: none; stroke: black; stroke-width: 0.2; }
    .text { font-family: DIN-Regular; font-size: 3.5px; fill: black; }
    .label { font-size: 2.8000000000000003px; }
  </style>
  <!-- Connectors -->
  <rect id="asm_power_2cond-endA" class="connector" x="60.00" y="136.00" width="20.00" height="15.00"/>
  <rect id="asm_power_2cond-endB" class="connector" x="340.00" y="136.00" width="20.00" height="15.00"/>
  <!-- Pin-1 Indicators -->
  <circle id="asm_power_2cond-pin1-L" cx="57.00" cy="138.00" r="1" fill="black"/>
  <circle id="asm_power_2cond-pin1-R" cx="363.00" cy="138.00" r="1" fill="black"/>
  <!-- Wires -->
  <path id="asm_power_2cond-net-+48V" class="wire" d="M 80.00,136.83 L 340.00,136.83" stroke="red"/>
  <path id="asm_power_2cond-net-RTN" class="wire" d="M 80.00,150.17 L 340.00,150.17" stroke="black"/>
  <!-- Dimensions -->
  <g id="asm_power_2cond-dim-oal" class="dimension">
    <line x1="60.00" y1="178.50" x2="360.00" y2="178.50"/>
    <line x1="60.00" y1="173.50" x2="60.00" y2="180.50"/>
    <line x1="360.00" y1="173.50" x2="360.00" y2="180.50"/>
    <text x="210.00" y="176.50" class="text" text-anchor="middle">800 ±10 mm</text>
  </g>
  <!-- Labels -->
  <text id="asm_power_2cond-label-0" x="60.00" y="126.00" class="text label">+48V POWER</text>
  <text id="asm_power_2cond-label-1" x="340.00" y="126.00" class="text label">LOAD</text>
  <!-- Notes -->
  <text x="20.00" y="287.00" class="text label">Notes: UL-LISTED</text>
  <!-- QR Code -->
  <rect id="asm_power_2cond-qr" x="390.00" y="267.00" width="20" height="20" fill="white" stroke="black" stroke-width="0.2"/>
  <text x="400.00" y="279.00" class="text label" text-anchor="middle" font-size="2">QR</text>
</svg>"
`;
